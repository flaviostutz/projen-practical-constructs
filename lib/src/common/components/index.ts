export * from './base-tooling';
export * from './common-target-type';
export * from './common-targets';
export * from './release-tasks';
export * from './publish';
