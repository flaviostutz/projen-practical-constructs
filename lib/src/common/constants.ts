export const PROJEN_VERSION = '0.91.13';
export const NODE_VERSION = '20.16.0';
