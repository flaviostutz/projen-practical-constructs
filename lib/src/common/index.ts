export * from './components';
export * from './files';
export * from './types';
