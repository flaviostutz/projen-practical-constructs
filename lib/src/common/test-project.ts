import { mkdtempSync } from 'node:fs';
import { tmpdir } from 'node:os';
// eslint-disable-next-line unicorn/import-style
import { join } from 'node:path';

import { Project } from 'projen';

export class TestProject extends Project {
  constructor() {
    const outdir = mkdtempSync(join(tmpdir(), 'test-'));
    super({ name: 'test', outdir });
  }
}
