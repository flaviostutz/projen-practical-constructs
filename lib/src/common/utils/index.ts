export * from './ts2jsii';
export * from './monotag';
