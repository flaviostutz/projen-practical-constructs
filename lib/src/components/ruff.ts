/* eslint-disable no-new */
import { Component, DependencyType, Project } from 'projen';

import { addTaskToParent, TaskOptions } from '../tasks';
import { RuffTomlFile, RuffTomlFileOptions } from '../files/ruff-toml';

export class Ruff extends Component {
  constructor(project: Project, opts: RuffOptions) {
    super(project);

    new RuffTomlFile(project, opts.ruffToml);
    project.deps.addDependency('ruff@0.8.*', DependencyType.DEVENV);
    project.addGitIgnore('.ruff_cache');
    const lintRuffTask = project.tasks.addTask('lint-ruff', {
      description: `Code checks (RUFF)`,
      steps: [
        {
          exec: `${opts.venvPath}/bin/ruff format --check src tests`,
        },
        {
          exec: `${opts.venvPath}/bin/ruff check src tests`,
        },
      ],
    });
    addTaskToParent(project, lintRuffTask, opts.attachTasksTo);
  }
}

export interface RuffOptions extends TaskOptions {
  /**
   * Ruff configuration file options
   */
  readonly ruffToml?: RuffTomlFileOptions;
}
