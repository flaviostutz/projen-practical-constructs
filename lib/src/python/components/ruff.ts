/* eslint-disable no-new */
import { Component, DependencyType, Project } from 'projen';

import { addSpawnTaskToExisting, TaskOptions } from '../tasks';
import { RuffTomlFile, RuffTomlFileOptions } from '../files/ruff-toml';
import { CommonTargets } from '../../common/components/common-target-type';

export class Ruff extends Component {
  constructor(project: Project, taskOpts: TaskOptions, opts?: RuffOptions) {
    super(project);

    new RuffTomlFile(project, opts);
    project.deps.addDependency('ruff@0.8.*', DependencyType.DEVENV);
    project.addGitIgnore('.ruff_cache');
    const lintRuffTask = project.tasks.addTask('lint-ruff', {
      description: `Code checks (RUFF)`,
      steps: [
        {
          exec: `${taskOpts.venvPath}/bin/ruff format --check src tests`,
        },
        {
          exec: `${taskOpts.venvPath}/bin/ruff check src tests`,
        },
      ],
    });
    addSpawnTaskToExisting(project, lintRuffTask, CommonTargets.LINT);

    const lintFixRuffTask = project.tasks.addTask('lint-ruff-fix', {
      description: `Lint fix (RUFF)`,
      steps: [
        {
          exec: `${taskOpts.venvPath}/bin/ruff format src tests`,
        },
        {
          exec: `${taskOpts.venvPath}/bin/ruff check --fix src tests`,
        },
      ],
    });
    addSpawnTaskToExisting(project, lintFixRuffTask, CommonTargets.LINT_FIX);
  }
}

export interface RuffOptions extends RuffTomlFileOptions {
  /**
   * Attach lint fix tasks to parent
   */
  readonly attachFixTaskTo?: string;
}
