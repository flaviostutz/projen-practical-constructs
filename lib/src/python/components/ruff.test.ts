/* eslint-disable no-new */
import { Testing } from 'projen';

import { TestProject } from '../../common/test-project';

import { Ruff } from './ruff';

describe('Ruff', () => {
  test('snapshot of default options', () => {
    const project = new TestProject();
    new Ruff(project, { venvPath: '.venv' });
    const out = Testing.synth(project);
    expect(out).toMatchSnapshot();
  });
  /* eslint-disable no-new */

  test('snapshot with custom options', () => {
    const project = new TestProject();
    new Ruff(project, { venvPath: '.venv' }, { attachFixTaskTo: 'pre-commit' });
    const out = Testing.synth(project);
    expect(out).toMatchSnapshot();
  });

  test('adds ruff dependency', () => {
    const project = new TestProject();
    new Ruff(project, { venvPath: '.venv' });
    Testing.synth(project);
    expect(project.deps.all.filter((d) => d.name === 'ruff').length).toBe(1);
  });

  test('adds ruff tasks', () => {
    const project = new TestProject();
    new Ruff(project, { venvPath: '.venv' });
    Testing.synth(project);
    expect(project.tasks.all.filter((t) => t.name === 'lint-ruff').length).toBe(1);
    expect(project.tasks.all.filter((t) => t.name === 'lint-ruff-fix').length).toBe(1);
  });
});
