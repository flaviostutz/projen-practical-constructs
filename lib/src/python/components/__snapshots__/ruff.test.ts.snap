// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ruff snapshot of default options 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/ruff.toml
.ruff_cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "ruff",
        "type": "devenv",
        "version": "0.8.*",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "ruff.toml",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "lint": {
        "name": "lint",
        "steps": [
          {
            "spawn": "lint-ruff",
          },
        ],
      },
      "lint-fix": {
        "name": "lint-fix",
        "steps": [
          {
            "spawn": "lint-ruff-fix",
          },
        ],
      },
      "lint-ruff": {
        "description": "Code checks (RUFF)",
        "name": "lint-ruff",
        "steps": [
          {
            "exec": ".venv/bin/ruff format --check src tests",
          },
          {
            "exec": ".venv/bin/ruff check src tests",
          },
        ],
      },
      "lint-ruff-fix": {
        "description": "Lint fix (RUFF)",
        "name": "lint-ruff-fix",
        "steps": [
          {
            "exec": ".venv/bin/ruff format src tests",
          },
          {
            "exec": ".venv/bin/ruff check --fix src tests",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "ruff.toml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

cache-dir = ".cache/.ruff_cache"
show-fixes = true
src = [ "src", "tests" ]
target-version = "py313"
unsafe-fixes = false

[format]
docstring-code-format = true

[lint]
ignore = [ ]
select = [ ]
per-file-ignores = { }

  [lint.mccabe]
  max-complexity = 14
",
}
`;

exports[`Ruff snapshot with custom options 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/ruff.toml
.ruff_cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "ruff",
        "type": "devenv",
        "version": "0.8.*",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "ruff.toml",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "lint": {
        "name": "lint",
        "steps": [
          {
            "spawn": "lint-ruff",
          },
        ],
      },
      "lint-fix": {
        "name": "lint-fix",
        "steps": [
          {
            "spawn": "lint-ruff-fix",
          },
        ],
      },
      "lint-ruff": {
        "description": "Code checks (RUFF)",
        "name": "lint-ruff",
        "steps": [
          {
            "exec": ".venv/bin/ruff format --check src tests",
          },
          {
            "exec": ".venv/bin/ruff check src tests",
          },
        ],
      },
      "lint-ruff-fix": {
        "description": "Lint fix (RUFF)",
        "name": "lint-ruff-fix",
        "steps": [
          {
            "exec": ".venv/bin/ruff format src tests",
          },
          {
            "exec": ".venv/bin/ruff check --fix src tests",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "ruff.toml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

cache-dir = ".cache/.ruff_cache"
show-fixes = true
src = [ "src", "tests" ]
target-version = "py313"
unsafe-fixes = false

[format]
docstring-code-format = true

[lint]
ignore = [ ]
select = [ ]
per-file-ignores = { }

  [lint.mccabe]
  max-complexity = 14
",
}
`;
